src/aiida/common/datastructures.py

```diff
@@ -217,3 +221,26 @@ class CodeRunMode(IntEnum):
 
     SERIAL = 0
     PARALLEL = 1
+
+
+class FileCopyOrder(IntEnum):
+    """Enum to indicate the order in which files are copied to the working directory.
+
+    There are three different sources of files that are copied to the working directory on the remote computer where a
+    calculation job is executed:
+
+        * Sandbox: files written to a temporary sandbox folder on the local file system written by the ``CalcJob``
+          plugin, first in the ``prepare_for_submission`` method, followed by the ``presubmit`` of the base class.
+        * Local: files written to the temporary sandbox folder by the engine based on the ``local_copy_list`` defined
+          by the plugin in the ``prepare_for_submission`` method.
+        * Remote: files written directly to the remote working directory by the engine base on the ``remote_copy_list``
+          defined by the plugin in the ``prepare_for_submission`` method.
+
+    Historically, these operations were performed in the order of local, sandbox and remote. For certain use cases,
+    however, this was deemed non-ideal, for example because files from the remote would override files written by the
+    plugin itself in the sandbox. This enum can be assigned to the ``CalcInfo`` instance returned by the calculation
+    job plugin ``prepare_for_submission`` implementation to change the order.
+    """
+
+    LOCAL_SANDBOX_REMOTE = 0
```

Comment:
I have one main question. Do we want to keep the combinatorial complexity of all possible permutations in a single enum? I know this is only 6 possibilities for now, but if we add more steps in the future
1. this enum becomes possibly complex
2. we need to properly support the old orders with only these three, the new ones, and the logic becomes complex

Would it be an option to have an enum defining the steps (LOCAL, SANDBOX, REMOTE) and then `calc_info.file_copy_order` is a list of those? (then validation can check unicity etc). 

For future backward compatibility, if not all are specified, anything not specified happens after those specified, in some predefined order (essentially you take the default order, pop those specified, and then do the rest at the end).

`None` would be equivalent to `[]` that would be equivalent to `[LOCAL, SANDBOX, REMOTE]`. 
The new order would be `[REMOTE, LOCAL, SANDBOX]`, that in principle (not recommended) one could also write simply as `[REMOTE]` (or `[REMOTE, LOCAL]`).